File: 06-DevOps/01-容器化/Docker层缓存失效/python/layer_cache.py
#!/usr/bin/env python3
"""Docker / BuildKit 构建缓存失效规则的最小模拟器。

规则依据 Docker 官方文档 "Build cache invalidation":
  https://docs.docker.com/build/cache/invalidation/

实现的三条核心规则:
  1. 缓存键 = 父层缓存键 + 本指令自身的摘要(逐层链式,父层变则后面全部失效)
  2. RUN 的摘要**只有命令字符串**,不检查容器文件系统
  3. COPY/ADD/RUN --mount=type=bind 的摘要 = 文件**元数据**校验和,
     且 **mtime 不参与**(只改 mtime 不会失效)

自检:python3 layer_cache.py   (全部断言必须通过)
"""
from __future__ import annotations

import hashlib
from dataclasses import dataclass, field

# 文件元数据 =(权限位, 内容, mtime)。mtime 显式建模进来,但**不参与校验和**
# —— 这样「只改 mtime 不失效」才是可断言的事实,而不是靠"模型里没有这个字段"。
Entry = tuple[int, str, float]
FileSet = dict[str, Entry]


def sha16(*parts: object) -> str:
    """把若干参数按固定分隔拼起来取 sha256 前 16 位十六进制。"""
    h = hashlib.sha256()
    for p in parts:
        h.update(str(p).encode())
        h.update(b"\x00")
    return h.hexdigest()[:16]


def metadata_checksum(files: FileSet) -> str:
    """算文件集合的元数据校验和。

    官方原文:"the builder calculates a cache checksum from file metadata ...
    The modification time of a file (mtime) is not taken into account when
    calculating the cache checksum."
    故此处只用 (路径, 权限位, 字节数, 内容哈希),**刻意丢弃 mtime**;
    且排序后遍历,保证结果与 readdir / dict 顺序无关。
    """
    h = hashlib.sha256()
    for path, (mode, content, _mtime) in sorted(files.items()):
        h.update(f"{path}\x00{mode:o}\x00{len(content)}\x00".encode())
        h.update(hashlib.sha256(content.encode()).hexdigest().encode())
        h.update(b"\x00")
    return h.hexdigest()[:16]


def restamp(files: FileSet, delta: float) -> FileSet:
    """把所有文件的 mtime 平移 delta 秒(模拟 git checkout / touch)。"""
    return {p: (m, c, t + delta) for p, (m, c, t) in files.items()}


# --------------------------------------------------------------------------
# 指令模型
# --------------------------------------------------------------------------
@dataclass
class Step:
    kind: str                 # FROM / WORKDIR / COPY / RUN / ARG
    text: str
    files: FileSet | None = None            # COPY / ADD 匹配到的文件
    secret: tuple[str, str] | None = None   # RUN --mount=type=secret=(id, 内容)

    def digest(self, args: dict[str, str], source_date_epoch: str) -> str:
        if self.kind in ("COPY", "ADD"):
            return sha16("copy", self.text, metadata_checksum(self.files or {}))
        if self.kind == "RUN" and self.secret:
            # 密钥**内容**不参与缓存,但 id / 挂载路径参与
            sid, _content = self.secret
            return sha16("run-secret", self.text, sid)
        if self.kind == "RUN":
            # 官方:"just the command string itself is used to find a match"
            return sha16("run", self.text)
        if self.kind == "WORKDIR":
            # 官方:WORKDIR 尊重 SOURCE_DATE_EPOCH,改它会级联失效
            return sha16("workdir", self.text, source_date_epoch)
        if self.kind == "ARG":
            name = self.text.split()[0]
            return sha16("arg", name, args.get(name, ""))
        return sha16(self.kind, self.text)


@dataclass
class BuildResult:
    hits: int = 0
    misses: int = 0
    run_misses: int = 0        # 代价高的 RUN 步骤被重跑的次数
    trace: list[str] = field(default_factory=list)

    def summary(self) -> str:
        return f"{self.hits} hit / {self.misses} miss (RUN 重跑 {self.run_misses})"


class CacheStore:
    """内容寻址的层缓存:key -> 是否已存在。"""

    def __init__(self) -> None:
        self.keys: set[str] = set()

    def lookup(self, key: str) -> bool:
        if key in self.keys:
            return True
        self.keys.add(key)
        return False


def build(steps: list[Step], store: CacheStore, args: dict[str, str] | None = None,
          source_date_epoch: str = "0") -> BuildResult:
    """模拟一次 docker build。

    父层规则:每一步的 key 都基于**上一步的 key**,所以父层一旦 miss,
    后面所有步骤的 key 都跟着变 —— 这就是「失效级联」,无需额外逻辑。
    """
    res = BuildResult()
    parent = sha16("scratch")
    for st in steps:
        key = sha16(parent, st.digest(args or {}, source_date_epoch))
        if store.lookup(key):
            res.hits += 1
            res.trace.append(f"CACHED  {st.kind:8} {st.text}")
        else:
            res.misses += 1
            if st.kind == "RUN":
                res.run_misses += 1
            res.trace.append(f"REBUILD {st.kind:8} {st.text}")
        parent = key
    return res


# --------------------------------------------------------------------------
# 两个 Dockerfile:坏顺序(先 COPY . . 再装依赖) vs 好顺序
# --------------------------------------------------------------------------
DEPS: FileSet = {"package.json": (0o644, '{"deps":["left-pad"],"mtime":[1700000000]}', 1.7e9)}


def app(content: str) -> FileSet:
    return {"app.py": (0o644, content, 1.7e9), "util.py": (0o644, "X = 1", 1.7e9)}


def bad_order(files: FileSet) -> list[Step]:
    """先把整个上下文 COPY 进去,再装依赖 —— 改一行源码就会重装依赖。"""
    merged: FileSet = {**DEPS, **files}
    return [
        Step("FROM", "python:3.12-slim"),
        Step("WORKDIR", "/app"),
        Step("COPY", "COPY . .", files=merged),
        Step("RUN", "pip install -r package.json"),
        Step("RUN", "python -m compileall ."),
    ]


def good_order(files: FileSet) -> list[Step]:
    """先 COPY 依赖清单装依赖,再 COPY 源码 —— 依赖层可复用。"""
    return [
        Step("FROM", "python:3.12-slim"),
        Step("WORKDIR", "/app"),
        Step("COPY", "COPY package.json ./", files=DEPS),
        Step("RUN", "pip install -r package.json"),
        Step("COPY", "COPY . .", files={**DEPS, **files}),
        Step("RUN", "python -m compileall ."),
    ]


# --------------------------------------------------------------------------
# 自检
# --------------------------------------------------------------------------
OK = 0
FAIL = 0


def check(label: str, cond: bool, detail: str = "") -> None:
    global OK, FAIL
    if cond:
        OK += 1
        print(f"  [PASS] {label} {detail}")
    else:
        FAIL += 1
        print(f"  [FAIL] {label} {detail}")


def main() -> int:
    v1, v2 = app("print('v1')"), app("print('v2')")

    print("=== 1. 冷启动:全部 MISS ===")
    store = CacheStore()
    cold = build(good_order(v1), store)
    check("冷启动无命中", cold.hits == 0 and cold.misses == 6, cold.summary())

    print("\n=== 2. 源码变更后的重建:好顺序 vs 坏顺序 ===")
    warm_good = build(good_order(v2), store)
    check("好顺序:前 4 步命中,COPY . . 起重建",
          warm_good.hits == 4 and warm_good.misses == 2, warm_good.summary())

    store_bad = CacheStore()
    build(bad_order(v1), store_bad)
    warm_bad = build(bad_order(v2), store_bad)
    check("坏顺序:COPY . . 起全部重建",
          warm_bad.hits == 2 and warm_bad.misses == 3, warm_bad.summary())
    # 公平口径:两个 Dockerfile 步数不同(5 vs 6),只比较**昂贵的 RUN 步骤**重跑次数
    check("好顺序保住依赖安装层",
          warm_bad.run_misses == 2 and warm_good.run_misses == 1,
          f"RUN 重跑 坏={warm_bad.run_misses} 好={warm_good.run_misses}")

    print("\n=== 3. mtime 不参与校验和 ===")
    check("只平移 mtime -> 校验和不变",
          metadata_checksum(v1) == metadata_checksum(restamp(v1, 86400 * 30)))
    check("只平移 mtime -> 校验和不变(含依赖清单)",
          metadata_checksum({**DEPS, **v1}) == metadata_checksum(restamp({**DEPS, **v1}, -1e6)))
    st_m = CacheStore()
    build(good_order(v1), st_m)
    after_touch = build(good_order(restamp(v1, 86400)), st_m)
    check("git checkout 改写全部 mtime -> 全部命中",
          after_touch.hits == 6 and after_touch.misses == 0, after_touch.summary())

    print("\n=== 4. chmod 改变元数据 -> COPY 失效级联 ===")
    chmodded = {"app.py": (0o755, "print('v1')", 1.7e9), "util.py": (0o644, "X = 1", 1.7e9)}
    check("内容相同但权限位不同 -> 校验和不同",
          metadata_checksum(v1) != metadata_checksum(chmodded))
    st_c = CacheStore()
    build(good_order(v1), st_c)
    after_chmod = build(good_order(chmodded), st_c)
    check("chmod 后从 COPY 起级联重建",
          after_chmod.hits == 4 and after_chmod.misses == 2, after_chmod.summary())

    print("\n=== 5. RUN 只看命令字符串,不看文件系统 ===")
    st_r = CacheStore()
    build([Step("FROM", "alpine:3.20"), Step("RUN", "apk add curl")], st_r)
    rerun = build([Step("FROM", "alpine:3.20"), Step("RUN", "apk add curl")], st_r)
    check("命令串未变 -> RUN 永远命中(包不会变新)",
          rerun.misses == 0 and rerun.hits == 2, rerun.summary())
    bumped = build([Step("FROM", "alpine:3.20"), Step("RUN", "apk add curl=8.9")], st_r)
    check("命令串变了才重建", bumped.misses == 1, bumped.summary())

    print("\n=== 6. ARG 参与缓存,构建密钥内容不参与 ===")
    arg_steps = [Step("ARG", "CACHEBUST = 1"), Step("RUN", "echo build")]
    st_a = CacheStore()
    build(arg_steps, st_a, args={"CACHEBUST": "1"})
    same = build(arg_steps, st_a, args={"CACHEBUST": "1"})
    diff = build(arg_steps, st_a, args={"CACHEBUST": "2"})
    check("ARG 值不变 -> 命中", same.misses == 0, same.summary())
    check("ARG 值改变 -> 级联失效", diff.hits == 0 and diff.misses == 2, diff.summary())

    st_s = CacheStore()
    build([Step("RUN", "some-command", secret=("TOKEN", "tkn_v1"))], st_s)
    rotated = build([Step("RUN", "some-command", secret=("TOKEN", "tkn_v2"))], st_s)
    check("轮换密钥内容 -> 仍命中", rotated.hits == 1 and rotated.misses == 0, rotated.summary())
    renamed = build([Step("RUN", "some-command", secret=("TOKEN2", "tkn_v1"))], st_s)
    check("改变密钥 id -> 失效", renamed.misses == 1, renamed.summary())

    print("\n=== 7. SOURCE_DATE_EPOCH 影响 WORKDIR 及其后续 ===")
    st_e = CacheStore()
    wd = [Step("WORKDIR", "/app"), Step("RUN", "make")]
    build(wd, st_e, source_date_epoch="0")
    ep = build(wd, st_e, source_date_epoch="1700000000")
    check("改 SOURCE_DATE_EPOCH -> WORKDIR 起失效",
          ep.hits == 0 and ep.misses == 2, ep.summary())

    print("\n=== 8. 失效级联不能跳步 ===")
    st_j = CacheStore()
    seq_v1 = [Step("FROM", "x"), Step("COPY", "COPY a", files={"a": (0o644, "1", 0.0)}),
              Step("RUN", "step-b"), Step("RUN", "step-c")]
    seq_v2 = [Step("FROM", "x"), Step("COPY", "COPY a", files={"a": (0o644, "2", 0.0)}),
              Step("RUN", "step-b"), Step("RUN", "step-c")]
    build(seq_v1, st_j)
    cas = build(seq_v2, st_j)
    check("首层变更 -> 后续全部重建,无法跳步复用",
          cas.hits == 1 and cas.misses == 3 and cas.run_misses == 2, cas.summary())

    print("\n=== 9. 指令顺序对缓存效率的量化影响 ===")
    # 同一份源码,只改 Dockerfile 行序:统计「改一行源码」的重建代价
    print(f"  坏顺序 {warm_bad.summary()}")
    print(f"  好顺序 {warm_good.summary()}")
    check("乱序步骤也应命中(理论上限)",
          build(list(reversed(good_order(v2))), CacheStore()).misses == 6)

    print(f"\n合计:{OK} passed / {FAIL} failed")
    return 1 if FAIL else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: 06-DevOps/01-容器化/Docker层缓存失效/go/layer_cache.go
// Docker / BuildKit 构建缓存失效规则的最小模拟器(Go 实现)。
//
// 规则依据 Docker 官方文档 "Build cache invalidation":
// https://docs.docker.com/build/cache/invalidation/
//
// 与 python/layer_cache.py 同题同算法,便于跨语言对照:
//   1. 缓存键 = 父层键 + 本指令摘要(hash 链,父层变则级联失效)
//   2. RUN 摘要只看命令字符串
//   3. COPY/ADD 摘要是文件"元数据"校验和,且 mtime 不参与
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sort"
	"strings"
)

// Entry 是一个文件的元数据:权限位、内容、修改时间。
// mtime 显式建模,但**不参与校验和** —— 这正是官方规则。
type Entry struct {
	Mode    uint32
	Content string
	Mtime   float64
}

type FileSet map[string]Entry

func sha16(parts ...string) string {
	h := sha256.New()
	for _, p := range parts {
		h.Write([]byte(p))
		h.Write([]byte{0})
	}
	return hex.EncodeToString(h.Sum(nil))[:16]
}

// MetadataChecksum 只用 (路径, 权限位, 字节数, 内容哈希),刻意丢弃 Mtime。
func MetadataChecksum(files FileSet) string {
	paths := make([]string, 0, len(files))
	for p := range files {
		paths = append(paths, p)
	}
	sort.Strings(paths)

	h := sha256.New()
	for _, p := range paths {
		e := files[p]
		body := sha256.Sum256([]byte(e.Content))
		h.Write([]byte(fmt.Sprintf("%s\x00%o\x00%d\x00", p, e.Mode, len(e.Content))))
		h.Write([]byte(hex.EncodeToString(body[:])))
		h.Write([]byte{0})
	}
	return hex.EncodeToString(h.Sum(nil))[:16]
}

// Step 是一条 Dockerfile 指令。
type Step struct {
	Kind   string // FROM / WORKDIR / COPY / RUN / ARG
	Text   string
	Files  FileSet
	Secret *SecretRef
}

// SecretRef 对应 RUN --mount=type=secret:内容不进缓存,ID 与挂载路径进缓存。
type SecretRef struct {
	ID      string
	Content string
}

func (s Step) digest(args map[string]string, sourceDateEpoch string) string {
	switch s.Kind {
	case "COPY", "ADD":
		return sha16("copy", s.Text, MetadataChecksum(s.Files))
	case "RUN":
		if s.Secret != nil {
			return sha16("run-secret", s.Text, s.Secret.ID)
		}
		// 官方:仅用命令字符串本身找匹配,不检查容器文件系统
		return sha16("run", s.Text)
	case "WORKDIR":
		return sha16("workdir", s.Text, sourceDateEpoch)
	case "ARG":
		name := strings.Fields(s.Text)[0]
		return sha16("arg", name, args[name])
	default:
		return sha16(s.Kind, s.Text)
	}
}

// CacheStore 是内容寻址的层缓存。
type CacheStore struct {
	keys map[string]bool
}

func NewCacheStore() *CacheStore { return &CacheStore{keys: map[string]bool{}} }

func (c *CacheStore) Lookup(key string) bool {
	if c.keys[key] {
		return true
	}
	c.keys[key] = true
	return false
}

// BuildResult 记录一次构建的命中情况。
type BuildResult struct {
	Hits, Misses, RunMisses int
	Trace                   []string
}

func (r BuildResult) Summary() string {
	return fmt.Sprintf("%d hit / %d miss (RUN 重跑 %d)", r.Hits, r.Misses, r.RunMisses)
}

// Build 模拟一次 docker build。每步的键基于上一步的键,故父层 miss 会级联。
func Build(steps []Step, store *CacheStore, args map[string]string, sourceDateEpoch string) BuildResult {
	var res BuildResult
	parent := sha16("scratch")
	for _, st := range steps {
		key := sha16(parent, st.digest(args, sourceDateEpoch))
		if store.Lookup(key) {
			res.Hits++
			res.Trace = append(res.Trace, fmt.Sprintf("CACHED  %-8s %s", st.Kind, st.Text))
		} else {
			res.Misses++
			if st.Kind == "RUN" {
				res.RunMisses++
			}
			res.Trace = append(res.Trace, fmt.Sprintf("REBUILD %-8s %s", st.Kind, st.Text))
		}
		parent = key
	}
	return res
}

var deps = FileSet{"package.json": {Mode: 0o644, Content: `{"deps":["left-pad"]}`, Mtime: 1.7e9}}

func app(content string) FileSet {
	return FileSet{
		"app.py":  {Mode: 0o644, Content: content, Mtime: 1.7e9},
		"util.py": {Mode: 0o644, Content: "X = 1", Mtime: 1.7e9},
	}
}

func merge(parts ...FileSet) FileSet {
	out := FileSet{}
	for _, p := range parts {
		for k, v := range p {
			out[k] = v
		}
	}
	return out
}

// BadOrder 与 GoodOrder 是同一份源码的两种 Dockerfile 写法。
func BadOrder(files FileSet) []Step {
	return []Step{
		{Kind: "FROM", Text: "python:3.12-slim"},
		{Kind: "WORKDIR", Text: "/app"},
		{Kind: "COPY", Text: "COPY . .", Files: merge(deps, files)},
		{Kind: "RUN", Text: "pip install -r package.json"},
		{Kind: "RUN", Text: "python -m compileall ."},
	}
}

func GoodOrder(files FileSet) []Step {
	return []Step{
		{Kind: "FROM", Text: "python:3.12-slim"},
		{Kind: "WORKDIR", Text: "/app"},
		{Kind: "COPY", Text: "COPY package.json ./", Files: deps},
		{Kind: "RUN", Text: "pip install -r package.json"},
		{Kind: "COPY", Text: "COPY . .", Files: merge(deps, files)},
		{Kind: "RUN", Text: "python -m compileall ."},
	}
}

func main() {
	fmt.Println("== Docker 层缓存失效规则模拟(Go) ==")

	v1, v2 := app("print('v1')"), app("print('v2')")

	// 场景 1:好顺序 —— 只改源码时依赖层被保住
	store := NewCacheStore()
	cold := Build(GoodOrder(v1), store, nil, "0")
	fmt.Printf("好顺序 冷启动   : %s\n", cold.Summary())
	warm := Build(GoodOrder(v2), store, nil, "0")
	fmt.Printf("好顺序 改源码后 : %s\n", warm.Summary())

	// 场景 2:坏顺序 —— 同一改动代价翻倍
	bad := NewCacheStore()
	Build(BadOrder(v1), bad, nil, "0")
	warmBad := Build(BadOrder(v2), bad, nil, "0")
	fmt.Printf("坏顺序 改源码后 : %s\n", warmBad.Summary())

	// 场景 3:mtime 不参与校验和
	restamped := FileSet{}
	for k, e := range v1 {
		e.Mtime += 86400 * 30
		restamped[k] = e
	}
	fmt.Printf("只平移 mtime    : 校验和相同=%v\n",
		MetadataChecksum(v1) == MetadataChecksum(restamped))

	// 场景 4:chmod 改元数据 -> COPY 及后续失效
	chmodded := FileSet{}
	for k, e := range v1 {
		chmodded[k] = e
	}
	e := chmodded["app.py"]
	e.Mode = 0o755
	chmodded["app.py"] = e
	fmt.Printf("chmod 后        : 校验和相同=%v\n",
		MetadataChecksum(v1) == MetadataChecksum(chmodded))

	// 场景 5:构建密钥轮换不失效
	secStore := NewCacheStore()
	Build([]Step{{Kind: "RUN", Text: "some-command", Secret: &SecretRef{"TOKEN", "tkn_v1"}}},
		secStore, nil, "0")
	rotated := Build([]Step{{Kind: "RUN", Text: "some-command", Secret: &SecretRef{"TOKEN", "tkn_v2"}}},
		secStore, nil, "0")
	renamed := Build([]Step{{Kind: "RUN", Text: "some-command", Secret: &SecretRef{"TOKEN2", "tkn_v1"}}},
		secStore, nil, "0")
	fmt.Printf("密钥内容轮换    : %s | 密钥 ID 改名: %s\n", rotated.Summary(), renamed.Summary())

	// 场景 6:ARG 参与缓存 -> 可用来强制失效
	argStore := NewCacheStore()
	steps := []Step{{Kind: "ARG", Text: "CACHEBUST = 1"}, {Kind: "RUN", Text: "echo build"}}
	Build(steps, argStore, map[string]string{"CACHEBUST": "1"}, "0")
	same := Build(steps, argStore, map[string]string{"CACHEBUST": "1"}, "0")
	diff := Build(steps, argStore, map[string]string{"CACHEBUST": "2"}, "0")
	fmt.Printf("ARG 相同        : %s | ARG 改变: %s\n", same.Summary(), diff.Summary())
}
